packages/narwallets/src/lib/narwallets.ts

```diff
@@ -0,0 +1,274 @@
+import { isMobile } from "is-mobile";
+import {
+  WalletModuleFactory,
+  WalletBehaviourFactory,
+  InjectedWallet,
+  Account,
+  Action,
+} from "@near-wallet-selector/core";
+import { providers } from "near-api-js";
+
+export interface SignAndSendTransactionParams {
+  signerId?: string;
+  receiverId?: string;
+  actions: Array<Action>;
+}
+export interface NarwalletsParams {
+  iconUrl?: string;
+}
+
+interface PendingPromises {
+  id: number;
+  code: string;
+  // Note: the result from different action may give different results, like boolean or FinalExecutionOutcome. Since the response may be many things, we're leaving any here
+  resolve: (value: any) => void;
+  reject: (reason?: any) => void;
+  timeout?: NodeJS.Timeout;
+}
+
+const NARWALLETS_CODES = {
+  SIGN_IN: "sign-in",
+  IS_INSTALLED: "is-installed",
+  IS_SIGNED_IN: "is-signed-in",
+  SIGN_OUT: "sign-out",
+  GET_ACCOUNT_ID: "get-account-id",
+  SIGN_AND_SEND_TRANSACTION: "sign-and-send-transaction",
+  SIGN_AND_SEND_TRANSACTIONS: "sign-and-send-transactions" 
+}
+
+let id = 0;
+const pendingPromises: Array<PendingPromises> = [];
+
+const isInstalled = async (): Promise<boolean> => {
+  // Note: sendToNarwallets throws if not installed
+  try {
+    await sendToNarwallets(NARWALLETS_CODES.IS_INSTALLED, true);
+  }
+  catch (ex) {
+    return false
+  }
+  return true;
+};
+
+const isSignedIn = (): Promise<boolean> => {
+  return sendToNarwallets(NARWALLETS_CODES.IS_SIGNED_IN, true);
+};
+
+const getAccountId = (): Promise<string> => {
+  return sendToNarwallets(NARWALLETS_CODES.GET_ACCOUNT_ID, true);
+};
+
+const callSignAndSendTransaction = (
+  params: SignAndSendTransactionParams
+): Promise<providers.FinalExecutionOutcome> => {
+  return sendToNarwallets(NARWALLETS_CODES.SIGN_AND_SEND_TRANSACTION, false, params);
+};
+
+const callSignAndSendTransactions = (
+  params: Array<SignAndSendTransactionParams>
+): Promise<
+  Array<providers.FinalExecutionOutcome>
+> => {
+  return sendToNarwallets(NARWALLETS_CODES.SIGN_AND_SEND_TRANSACTIONS, false, params);
+};
+
+const sendToNarwallets = (
```

Comment:
This function is being used above before it's been declared which is not good practice and it's surprising that it's working.

Can you move it up to line 41?

Reply:
Although this approach seemed to work here but if you try to use an arrow function before it's declared JS would throw a syntax error.

So this would normally throw this error: **Uncaught ReferenceError: Cannot access 'someFunctionA' before initialization**


```ts
someFunctionA();

const someFunctionA = () => {
  console.log("Hello World!");
}
```

But if you declare the function this way it would not throw in this case:

```ts
  someFunctionA();

  function someFunctionA() {
    console.log("Hello World!");
  }
  ```
  
  This is due to how [Hoisting ](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting)in JS works